src/coreclr/jit/earlyprop.cpp

```diff
@@ -645,15 +648,35 @@ void Compiler::optFoldNullCheck(GenTree* tree, LocalNumberToNullCheckTreeMap* nu
 //       or
 //       indir(add(x, const2))
 //
-//       (indir is any node for which OperIsIndirOrArrLength() is true.)
+//       (indir is any node for which OperIsIndirOrArrLength() or OperIsImplicitIndir() is true.)
 //
 //     2.  const1 + const2 if sufficiently small.
-
+//
 GenTree* Compiler::optFindNullCheckToFold(GenTree* tree, LocalNumberToNullCheckTreeMap* nullCheckMap)
 {
-    assert(tree->OperIsIndirOrArrLength());
+    assert(tree->OperIsIndirOrArrLength() || tree->OperIsImplicitIndir());
 
-    GenTree* addr = (tree->OperGet() == GT_ARR_LENGTH) ? tree->AsArrLen()->ArrRef() : tree->AsIndir()->Addr();
+    GenTree* addr;
+    if (tree->OperIs(GT_ARR_LENGTH))
+    {
+        addr = tree->AsArrLen()->ArrRef();
+    }
+    else if (tree->OperIsIndir())
+    {
+        addr = tree->AsIndir()->Addr();
+    }
+    else if (tree->OperIs(GT_XORR, GT_XAND, GT_XADD, GT_XCHG))
+    {
+        const bool commaOnly = true;
+        addr                 = tree->AsOp()->gtGetOp1()->gtEffectiveVal(commaOnly);
+    }
+    else
+    {
+        // TODO-CQ: Support other implicit indir cases.
```

Comment:
There are other instructions in `OperIsImplicitIndir` that are not supported here, like `GT_BOX`, `AsHWIntrinsic()->OperIsMemoryLoadOrStore()` or `AsSIMD()->OperIsMemoryLoad()`. I have not added them because was not sure how to access the address node in the expected format.

Reply:
You can at least add `GT_LOCKADD`, `GT_CMPXCHG`, and `GT_ARR_INDEX`. All of them have the address in `gtOp1`. For `GT_ARR_ELEM` and `GT_ARR_OFFSET` you'll need to access `gtArrObj`.` GT_BOX` is not interesting here, it's expanded early and we'll see explicit `GT_IND` nodes.

Reply:
We can have `GT_BOX` here, saw some asserts during SPC crossgen, is it unexpected?
[unexpectedBoxInEarlyProp.txt](https://github.com/dotnet/runtime/files/6246923/unexpectedBoxInEarlyProp.txt)
I have attached JitDump where BOX survives until rationalize.

Reply:
Yes, you can see `BOX` instructions at that point but they no longer have an implicit indirection. We fully expand boxes in `impImportAndPushBox` but we also leave `BOX` marker instructions to help some optimizations. For example, in your dump we have

```
***** BB31
STMT00045 (IL 0x196...0x1AE)
               [000216] -A----------              *  ASG       ref   
               [000215] D------N----              +--*  LCL_VAR   ref    V20 tmp4         
               [000214] ------------              \--*  ALLOCOBJ  ref   
               [000213] #-----------                 \--*  IND       long  
               [000212] H-----------                    \--*  CNS_INT(h) long   0x4000000000420358 class

***** BB31
STMT00046 (IL   ???...  ???)
               [000221] -A--G-------              *  ASG       struct (copy)
               [000220] ------------              +--*  BLK       struct<System.Guid, 16>
               [000219] ------------              |  \--*  ADD       byref 
               [000217] ------------              |     +--*  LCL_VAR   ref    V20 tmp4         
               [000218] ------------              |     \--*  CNS_INT   long   8
               [000211] ----G-------              \--*  FIELD     struct ClassId
               [000210] ------------                 \--*  ADDR      byref 
               [000209] -------N----                    \--*  LCL_VAR   struct<Internal.Runtime.InteropServices.ComActivationContext, 56> V00 arg0         

***** BB31
STMT00047 (IL   ???...  ???)
               [000224] I-C-G-------              *  CALL r2r_ind ref    System.String.Format (exactContextHnd=0x4000000000420039)
               [000208] ------------ arg0         +--*  CNS_STR   ref   <string constant>
               [000223] ------------ arg1         \--*  BOX       ref   
               [000222] ------------                 \--*  LCL_VAR   ref    V20 tmp4 
```

As you can see the interesting indirection is no longer part of the `BOX`, it's

```
               [000211] ----G-------              \--*  FIELD     struct ClassId
               [000210] ------------                 \--*  ADDR      byref 
               [000209] -------N----                    \--*  LCL_VAR   struct<Internal.Runtime.InteropServices.ComActivationContext, 56> V00 arg0 
```